File: 20. Valid Parentheses Easy/solution.py
def is_valid(s: str) -> bool:
    BRACKETS = {
        ")": "(",
        "]": "[",
        "}": "{",
    }
    result = []

    for bracket in s:
        if bracket in BRACKETS:
            top_char = result.pop() if result else "#"
            if BRACKETS[bracket] != top_char:
                return False
        else:
            result.append(bracket)
    return not result

print(is_valid("()"))
print(is_valid("()[]{}"))
print(is_valid("(]"))

File: 20. Valid Parentheses Easy/solution.go
package main

import "fmt"

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
}

func isValid(s string) bool {
	BRACKETS := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	result := []rune{}

	for _, bracket := range s {
		if bracketIsOpen(bracket) {
			result = append(result, bracket)
		} else {
			if len(result) == 0 || result[len(result)-1] != BRACKETS[bracket] {
				return false
			}
			result = result[:len(result)-1]
		}
	}
	return len(result) == 0
}

func bracketIsOpen(char rune) bool {
	return char == '(' || char == '{' || char == '['
}
